File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def DFS(node, maximum):
            ret = 0
            if node.left:
                ret += DFS(node.left, max(maximum, node.val)) 
            if node.right:
                ret += DFS(node.right, max(maximum, node.val)) 
            if maximum <= node.val:
                ret += 1
            return ret
        return DFS(root, -10**4)
            

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {
        int ans = 0;
        DFS(root, root->val, ans); 
        return ans;
    }
private:
    void DFS(TreeNode* node, int maxNum, int& ans) {
        if (node == nullptr) return;
        if (node->val >= maxNum) ans++;
        DFS(node->left, max(maxNum, node->val), ans);
        DFS(node->right, max(maxNum, node->val), ans);
    }
};